### ISSUE TITLE
[CI/CD] Compilation issue for Arch Linux Usan 3.x/4532 

### ISSUE THREAD

**Issue by picnixz** [python/cpython#122523]
# Bug report

### Bug description:

Tests are broken for build bot Arch Linux Usan 3.x/4532 since the compilation fails with:

```
Objects/object.c:1071:16: runtime error: call to function long_hash through pointer to incorrect function type 'long (*)(struct _object *)'
/buildbot/buildarea/3.x.pablogsal-arch-x86_64.clang-ubsan/build/Objects/longobject.c:3618: note: long_hash defined here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior Objects/object.c:1071:16 in
make: *** [Makefile:1691: Python/frozen_modules/getpath.h] Error 1
```

See for instance https://buildbot.python.org/#/builders/720/builds/1300.

### CPython versions tested on:

CPython main branch

**Comment by picnixz** [comment]
Because of the sheer amount of work to do (I just looked at what I needed to do, fixed 1-2 things and then gave up), I suggest adding a new build bot with `-fsanitize=undefined -fno-sanitize=function -fsanitize-recover`. That way, we could catch the other undefined behaviours that are not due to those casts. Currently, the build bot fails fast (and we only see 1 UB per build due to the `-fno-sanitize-recover`) but I think that most of the time, we won't have a lot of UBs (or else it's bad for us...). As such, having a build bot that builds entirely (with `-fsanitize-recover`) instead of fast-failing would help a lot in detecting the other cases.

Once this new build bot is green, we can focus on eliminating the UBs related to function pointers.

Note: see https://github.com/python/cpython/pull/123004#issuecomment-2288611116 for what was detected as a first pass. I suspect there are UBs that are not detected but that's for another day.

_Originally posted by @picnixz in https://github.com/python/cpython/issues/111178#issuecomment-2289087204_

**Comment by picnixz** [comment]
@pablogsal I was told that you were responsible for the USAN build bot. Could you consider:
adding a build bot with the above configuration (namely, `-fsanitize=undefined -fno-sanitize=function -fsanitize-recover`) please?

**Comment by erlend-aasland** [comment]
Should this issue be moved to the buildbot repo?

**Comment by picnixz** [comment]
Oh yes, please. How can I do it easily? Should I just close this one and open a new one out there? (I forgot about the other repository...)

**Comment by picnixz** [comment]
> Should this issue be moved to the buildbot repo?

Is the repository in question: https://github.com/python/buildmaster-config? or is it another one?

**Comment by pablogsal** [comment]
> @pablogsal I was told that you were responsible for the USAN build bot. Could you consider
> 
> adding a build bot with the above configuration (namely, `-fsanitize=undefined -fno-sanitize=function -fsanitize-recover`) please?

Yeah absolutely! Do you want another bot with those settings or to just substitute the current one?
